company-lsp.el

```diff
@@ -105,6 +105,48 @@ CANDIDATE is a string returned by `company-lsp--make-candidate'."
       (delete-char (length label))
       (insert insert-text))))
 
+(defvar-local company-lsp--pending-requests (make-hash-table))
+
+(defcustom company-lsp-async nil
+  "Whether or not to use async operations to fetch data."
+  :type 'boolean
+  :group 'company-lsp)
+
+(defun company-lsp--on-message (p msg)
+  "Function called just after `lsp--parser-on-message' using `advice-add'.
+This hook processes all the responses and filter the ones with an id that is in
+`company-lsp-ids'.
+P and MSG are the parameters from `lsp--parser-on-message'."
+  (let* ((json-array-type 'list)
+	 (json-object-type 'hash-table)
+	 (json-false nil)
+	 (json-data (json-read-from-string msg)))
+    (pcase (lsp--get-message-type json-data)
+      ('response
+       (when-let* ((id-msg (gethash "id" json-data nil))
+		   (callback (gethash id-msg company-lsp--pending-requests)))
+	 (remhash id-msg company-lsp--pending-requests)
+	 (let* ((response (lsp--parser-response-result p))
+		(items (cond ((hash-table-p response) (gethash "items" response nil))
```

Comment:
Can you extract the (let ((items ...)) (funcall ...)) into a function company-lsp--on-completion (response company-callbacl)? Both async and sync completion use the same logic.